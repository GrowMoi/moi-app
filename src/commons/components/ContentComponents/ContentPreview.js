import React, { Component } from 'react';
import { Ionicons } from '@expo/vector-icons';
import PropTypes from 'prop-types';
import { View, Image, TouchableOpacity, TouchableWithoutFeedback } from 'react-native';
import * as Animatable from 'react-native-animatable';
import styled from 'styled-components/native';
import { TextBody, Header } from '../Typography';
import { Size, Palette } from '../../styles';
import learntCheck from '../../../../assets/images/miscelaneous/check.png';

const RowContainer = styled(View)`
  flex-direction: row;
  align-self: stretch;
  background-color: ${(props) => {
    if (props.inverted) return Palette.white.alpha(0.2).css();
    return 'transparent';
  }};
  border-radius: 10;
  shadow-color: ${Palette.dark};
  shadow-opacity: 0.2;
  shadow-offset: -5px 5px;
  shadow-radius: 2;
  margin-bottom: ${Size.spaceSmall};
  padding-horizontal: ${Size.spaceSmall};
  padding-vertical: ${Size.spaceSmall};
  border-color: ${(props) => {
    if (props.inverted) return 'transparent';
    return Palette.white.alpha(0.2).css();
  }};
  border-width: 1;
`;

const imageWidth = 150;
const imageHeight = 120;
const aspect = imageWidth / imageHeight;
const ContentImage = styled(Image)`
  width: ${props => props.width};
  height: ${props => Math.round(props.width / aspect)};
  background-color: ${Palette.dark};
`;

const Content = styled(View)`
  flex: 1;
  padding-horizontal: ${Size.spaceSmall};
  padding-vertical: ${Size.spaceXSmall};
  position: relative;
`;

const CloseAction = styled(TouchableOpacity)`
  position: absolute;
  top: -5;
  right: -10;
  width: 20;
  height: 20;
  border-radius: 10;
  background-color: ${Palette.dark};
  align-items: center;
  justify-content: center;
`;

const LearntImage = styled(Image)`
  position: absolute;
  top: -10;
  right: -5;
  width: 50;
  height: 50;
`

const LearntOverlay = styled(View)`
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background-color: rgba(0,0,0,0.5);
  border-radius: 10;
`

export default class ContentPreview extends Component {
  static propTypes = {
    inverted: PropTypes.bool,
    source: PropTypes.any,
    title: PropTypes.string,
    subtitle: PropTypes.string,
    description: PropTypes.string,
    onPress: PropTypes.func,
    id: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string,
    ]),
    width: PropTypes.number,
    closeButton: PropTypes.bool,
    onPressCloseButton: PropTypes.func,
    learnt: PropTypes.bool,
  }

  static defaultProps = {
    width: 100,
  }

  handleViewRef = ref => this.contentRow = ref;

  rowContent = () => {
    const { inverted, title, subtitle, description, source, width, closeButton } = this.props;

    const content = (
      <Content key='row-description'>
        <Header ellipsizeMode='tail' numberOfLines={3} customSize={18} bolder>{title}</Header>
        <TextBody small secondary>{subtitle}</TextBody>
        <TextBody small ellipsizeMode='tail' numberOfLines={4}>{description}</TextBody>
      </Content>
    );

    const image = (
      <ContentImage
        width={width}
        key='row-image'
        resizeMode='cover'
        source={source} />
    );

    const contentOrder = inverted ? [content, image] : [image, content];
    return contentOrder;
  }

  onPressRow = () => {
    const { onPress } = this.props;
    if(onPress) {
      this.contentRow.pulse(300)
        .then(endState => {
          if(endState.finished) {
            onPress();
          }
        })
    }
  }

  render() {
    const { inverted, id, closeButton, onPressCloseButton, animationDelay = 0, withoutAnimation = false, learnt = false, onPressIn = ()=>{} } = this.props;
    const closeAction = closeButton && (
      <CloseAction onPress={() => onPressCloseButton && onPressCloseButton(id)}>
        <Ionicons name='md-close' size={15} color={Palette.white.css()}/>
      </CloseAction>
    );

    const content = (
      <TouchableWithoutFeedback onPress={this.onPressRow} onPressIn={onPressIn}>
        <Animatable.View ref={this.handleViewRef} style={{ alignSelf: 'stretch' }}>
          <RowContainer id={id} inverted={inverted}>
            {closeAction}
            {this.rowContent()}
            {learnt && (
              <LearntOverlay>
                <LearntImage source={learntCheck} resizeMode='contain' />
              </LearntOverlay>
            )}
          </RowContainer>
        </Animatable.View>
      </TouchableWithoutFeedback>
    );

    if(withoutAnimation) return content;
    return (
      <Animatable.View
        animation="bounceIn"
        delay={animationDelay}
      >
        {content}
      </Animatable.View>
    );
  }
}
