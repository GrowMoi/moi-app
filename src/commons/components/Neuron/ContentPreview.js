import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { View, Image, TouchableOpacity } from 'react-native';
import styled from 'styled-components/native';
import { TextBody, Header } from '../Typography';
import { Size, Palette } from '../../styles';

const RowContainer = styled(TouchableOpacity)`
  flex-direction: row;
  align-self: stretch;
  background-color: ${Palette.white};
  border-color: ${Palette.neutral.alpha(0.2).css()};
  border-width: 1;
`;

const ContentImage = styled(Image)`
  width: 150;
  height: 120;
  background-color: ${Palette.dark};
`;

const Content = styled(View)`
  flex: 1;
  padding-horizontal: ${Size.spaceSmall};
  padding-vertical: ${Size.spaceXSmall};
`;

export default class ContentPreview extends Component {
  static propTypes = {
    inverted: PropTypes.bool,
    source: PropTypes.any,
    title: PropTypes.string,
    subtitle: PropTypes.string,
    description: PropTypes.string,
    onPress: PropTypes.func,
  }

  get rowContent() {
    const { inverted, title, subtitle, description, source } = this.props;

    const content = (
      <Content key='row-description'>
        <Header ellipsizeMode='tail' numberOfLines={1}>{title}</Header>
        <TextBody small secondary>{subtitle}</TextBody>
        <TextBody small ellipsizeMode='tail' numberOfLines={4}>{description}</TextBody>
      </Content>
    );

    const image = (
      <ContentImage
        key='row-image'
        resizeMode='cover'
        source={source} />
    );

    if (inverted) return [content, image];
    return [image, content];
  }

  render() {
    const { onPress } = this.props;
    return (
      <RowContainer onPress={onPress}>
        {this.rowContent}
      </RowContainer>
    );
  }
}
