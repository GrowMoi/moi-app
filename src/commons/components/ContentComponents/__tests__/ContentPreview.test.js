import React from 'react';
import 'react-native';
import { shallow } from 'enzyme';
import ContentPreview from '../ContentPreview';

describe('Content Preview', () => {
  it('it should render row as expected', () => {
    const wrapper = shallow(
      <ContentPreview
        source={{ uri: 'http://www.wolaver.org/space/crab.jpg' }}
        title='Grab Nebula'
        subtitle='Que es?'
        description='lorem lorem loremlorem lorem lorem laskdj flakd jfalskdfj alskdf jsadkfj jasdl;fk jasdfkl diasdj asdfh aksdfjh akdfjh akldfjhas fjd'
      />,
    );
    const tree = wrapper.dive();

    expect(tree).toMatchSnapshot();
    expect(tree).toBeTruthy();
  });
});
