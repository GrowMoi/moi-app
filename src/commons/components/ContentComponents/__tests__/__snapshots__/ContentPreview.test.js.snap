// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Preview it should render row as expected 1`] = `
<withAnimatable(View)
  accessible={true}
  delay={0}
  direction="normal"
  iterationCount={1}
  onAnimationBegin={[Function]}
  onAnimationEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignSelf": "stretch",
    }
  }
  useNativeDriver={false}
>
  <Styled(View)>
    <Styled(Image)
      resizeMode="cover"
      source={
        Object {
          "uri": "http://www.wolaver.org/space/crab.jpg",
        }
      }
      width={100}
    />
    <Styled(View)>
      <Styled(Text)
        ellipsizeMode="tail"
        numberOfLines={2}
      >
        Grab Nebula
      </Styled(Text)>
      <Styled(Text)
        secondary={true}
        small={true}
      >
        Que es?
      </Styled(Text)>
      <Styled(Text)
        ellipsizeMode="tail"
        numberOfLines={4}
        small={true}
      >
        lorem lorem loremlorem lorem lorem laskdj flakd jfalskdfj alskdf jsadkfj jasdl;fk jasdfkl diasdj asdfh aksdfjh akdfjh akldfjhas fjd
      </Styled(Text)>
    </Styled(View)>
  </Styled(View)>
</withAnimatable(View)>
`;
